nx/lib/nx.ex

```diff
@@ -12905,21 +12966,60 @@ defmodule Nx do
                 "expected start and stop to have the same shape. Got shapes #{inspect(start_shape)} and #{inspect(stop_shape)}"
       end
 
+    {_start, _stop, vectorized_axes} = devectorize_with_axes(start, stop)
+
     divisor =
       if opts[:endpoint] do
         n - 1
       else
         n
       end
 
+    iota = iota(iota_shape, axis: -1, type: opts[:type], vectorized_axes: vectorized_axes)
+
     step = Nx.subtract(stop, start) |> Nx.divide(divisor)
 
-    iota = iota(iota_shape, axis: -1, type: opts[:type])
+    out =
+      iota
+      |> multiply(step)
+      |> add(start)
+      |> as_type(opts[:type])
+
+    cond do
+      out.vectorized_axes == [] ->
+        out
 
-    iota
-    |> multiply(step)
-    |> add(start)
-    |> as_type(opts[:type])
+      out.vectorized_axes == vectorized_axes ->
+        out
+
+      true ->
+        nil
+        # we need to reorder the vectorized axes to the expected order
+
+        idx_by_name = Enum.with_index(vectorized_axes, fn {name, _}, idx -> {name, idx} end)
+
+        {[], indices_rev} =
+          Enum.reduce(out.vectorized_axes, {idx_by_name, []}, fn {name, _},
+                                                                 {idx_by_name, indices} ->
+            {{_name, idx}, idx_by_name} = List.keytake(idx_by_name, name, 0)
+
+            unless idx do
+              raise "invalid vectorized axes"
+            end
+
+            {idx_by_name, [idx | indices]}
+          end)
+
+        devec = devectorize(out)
+
+        inner_axes = count_up(tuple_size(out.shape), tuple_size(devec.shape) - 1)
+
+        transpose_axes = Enum.reverse(indices_rev) ++ inner_axes
+
+        devec
+        |> transpose(axes: transpose_axes)
+        |> revectorize_and_validate_sizes(vectorized_axes)
```

Comment:
This here could be extracted to `transpose_vectorized_axes` or something like that, that acts like transpose, but _only_ on vectorized axes.

Reply:
@josevalim I tested this locally with the changes in #1136 

The only changes to linspace is that we would call `reshape_vectors` on `start` and `stop` instead of the `to_tensor` calls and we extract vectorized_axes from either to build the iota via the `:vectorized_axes` option.

We don't even need `broadcast_vectors` because the other operations already end up taking care of this implicitly for us!